shell/platform/windows/flutter_windows_engine_unittests.cc

```diff
@@ -619,5 +620,142 @@ TEST_F(FlutterWindowsEngineTest, AlertPlatformMessage) {
   }
 }
 
+class MockWindowsLifecycleManager : public WindowsLifecycleManager {
+ public:
+  MockWindowsLifecycleManager(FlutterWindowsEngine* engine)
+      : WindowsLifecycleManager(engine) {}
+  virtual ~MockWindowsLifecycleManager() {}
+
+  MOCK_METHOD2(Quit, void(UINT, HWND));
+};
+
+TEST_F(FlutterWindowsEngineTest, TestExit) {
+  FlutterWindowsEngineBuilder builder{GetContext()};
+  builder.SetDartEntrypoint("exitTestExit");
+  bool finished = false;
+  bool did_call = false;
+
+  std::unique_ptr<FlutterWindowsEngine> engine = builder.Build();
+
+  EngineModifier modifier(engine.get());
+  modifier.embedder_api().RunsAOTCompiledDartCode = []() { return false; };
+  auto handler = std::make_unique<MockWindowsLifecycleManager>(engine.get());
+  ON_CALL(*handler, Quit).WillByDefault([&finished](UINT exit_code, HWND hwnd) {
+    finished = exit_code == 0;
+  });
+  EXPECT_CALL(*handler, Quit).Times(1);
+  modifier.SetLifecycleManager(std::move(handler));
+
+  auto binary_messenger =
+      std::make_unique<BinaryMessengerImpl>(engine->messenger());
+  binary_messenger->SetMessageHandler(
+      "flutter/platform", [&did_call](const uint8_t* message,
+                                      size_t message_size, BinaryReply reply) {
```

Comment:
Can we check the message's content?

Reply:
We can, albeit perhaps a bit clunkily. What do you think we should check it for?

Reply:
The concern would be that we accidentally regress the messages' contents.

Why would it be clunky? Can you not cast it to a `std::string` and then use `EXPECT_EQ`? For example:

https://github.com/flutter/engine/blob/252f6cfd0335918d4147ba63e52b5800c7254478/shell/platform/windows/platform_handler_unittests.cc#L113

https://github.com/flutter/engine/blob/252f6cfd0335918d4147ba63e52b5800c7254478/shell/platform/windows/platform_handler_unittests.cc#L200-L202

Reply:
Yes, I was just referring to expecting the json message object to be equal to the string. I think we can know the order, etc., ahead of time, but it might not be obvious, since multiple valid strings would represent the correct message.

Reply:
`The plain [HashMap](https://api.dart.dev/be/180791/dart-collection/HashMap-class.html) is unordered (no order is guaranteed),`
This leads be to believe that the `jsonString` created at https://github.com/flutter/engine/pull/40802/files/ec79b21b9a93211115228da555d0d02a5c227287#diff-0ec6bbe0e6c9b89b68745c51c5e6beda1692354554e97b6c1789bb4614ae8fdfR103 and below does not have a guaranteed order, so testing for equality with a string literal would not be as simple as testing the list example you provided, which is ordered.

Reply:
Hm that's a fair concern. We have many tests that assume the order is stable. If Dart ever changes how hash map's order is defined, our tests that use string equality checks on JSON will break and we'll need to fix them.

Regardless, we should verify the message is expected. Feel free to choose whichever approach you prefer :)

Reply:
Dart's default Map class is not a HashMap but a LinkedHashMap, which guarantees order.